### PR TITLE
AutoYaST encryption support

### PR DESCRIPTION
**Pull request by imobachgs**
This PR extends AutoYaST to support the new encryption methods. Basically, it adds a new `crypt_method` element which allows to set the encryption method to use (`luks1`, `pervasive_luks2`, `secure_swap`, `random_swap` and `protected_swap`).

```xml
   <drive>
    <type config:type="symbol">CT_DISK</type>
    <use>all</use>
    <partitions config:type="list">
     <partition>
      <size>20G</size>
      <mount>/</mount>
      <filesystem config:type="symbol">ext4</filesystem>
      <crypt_method config:type="symbol">luks1</crypt_method> <!-- default method if crypt_key is defined -->
      <crypt_key>S3CR3T</crypt_key>
     </partition>
     <partition>
      <size>1G</size>
      <mount>swap</mount>
      <crypt_method config:type="symbol">random_swap</crypt_method> <!-- set encryption method -->
     </partition>
    </partitions>
   </drive>
```

## Deprecating `crypt_fs`

Setting `crypt_fs` to `true` is equivalent to setting `crypt_method` to `luks1`. However, `crypt_fs` is deprecated and it is not exported when cloning the system.

## User feedback

AutoYaST tries to help the user by detecting and reporting these issues:

* The encryption method is unknown.
* The encryption method is not available (for instance, `pervasive_luks2` in an x86 system).
* The encryption method is not suitable for the current device (e.g., trying to cipher `/home` with `random_swap`).

## Screenshots

`random_swap` is only suitable for swap partitions

`pervasive_luks2` is not available in non s390 systems

## Pending tasks

These tasks are still pending, but they have to be addressed in different repositories:

### CODE REVIEW

File: src/lib/y2storage/autoinst_issues/invalid_encryption.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    # @example The encryption method is not suitable for the device
+    #   hash = { "mount" => "/", "crypt_method" => :random_swap }
+    #   section = AutoinstProfile::PartitionSection.new_from_hashes(hash)
+    #   issue = InvalidEncryption.new(section, :unsuitable)
+    class InvalidEncryption < Issue
+      # @return [Symbol] Reason which causes the encryption to be invalid
+      attr_reader :reason
+
+      # @param section [#parent,#section_name] Section where it was detected (see {AutoinstProfile})
+      # @param reason  [:unknown, unavailable] Reason which casues the encryption to be invalid
```
Comment (dgdavid) [inline src/lib/y2storage/autoinst_issues/invalid_encryption.rb]: Missing colon, `:unavailable`

Comment (dgdavid) [reply src/lib/y2storage/autoinst_issues/invalid_encryption.rb]: Also missing the `:unsuitable` possibility

File: src/lib/y2storage/proposal/autoinst_drive_planner.rb [hunk 2]
```diff
@@ -82,12 +81,58 @@ def configure_filesystem(device, partition_section, drive_section)
         configure_subvolumes(device, partition_section)
       end
 
+      DEFAULT_ENCRYPTION_METHOD = EncryptionMethod.find(:luks1)
+      private_constant :DEFAULT_ENCRYPTION_METHOD
+
       # Sets common devices attributes
```
Comment (dgdavid) [inline src/lib/y2storage/proposal/autoinst_drive_planner.rb]: s/devices/encryption ?

File: src/lib/y2storage/proposal/autoinst_drive_planner.rb [hunk 3]
```diff
@@ -82,12 +81,58 @@ def configure_filesystem(device, partition_section, drive_section)
...
+        error =
+          if encryption_method.nil?
+            :unknown
+          elsif !encryption_method.available?
+            :unavailable
+          elsif !device.supported_encryption_method?(encryption_method)
+            :unsuitable
+          end
+
+        issues_list.add(:invalid_encryption, partition_section, error) if error
```
Comment (dgdavid) [inline src/lib/y2storage/proposal/autoinst_drive_planner.rb]: Why not only one `error` check?. I mean, 

```ruby
if error
  issues_list.add(:invalid_encryption, partition_section, error)

  return
end

encryption_method
```

or 

```ruby
if error
  issues_list.add(:invalid_encryption, partition_section, error)
  nil
else
  encryption_method
end 
```

File: test/y2storage/autoinst_profile/partition_section_test.rb [hunk 4]
```diff
@@ -590,15 +590,29 @@ def section_for(name)
     end
 
     context "if the partition is encrypted" do
```
[the remark this answers is not in the document]
Comment (imobachgs) [reply test/y2storage/autoinst_profile/partition_section_test.rb]: OK, you are right. I've updated the PR to just ignore the encryption settings in such a case.

File: test/y2storage/autoinst_profile/partition_section_test.rb [hunk 5]
```diff
@@ -590,15 +590,35 @@ def section_for(name)
...
-      it "initializes #crypt_key to a generic string" do
-        expect(section_for("sdf7").crypt_key).to eq "ENTER KEY HERE"
+      context "and the encryption method requires a password" do
+        it "initializes #crypt_key to a generic string" do
+          expect(section_for("sdf7").crypt_key).to eq "ENTER KEY HERE"
+        end
       end
 
-      it "initializes #loop_fs and #crypt_key to true" do
+      context "and the encryption method is unknown" do
```
[the remark this answers is not in the document]
Comment (ancorgs) [reply test/y2storage/autoinst_profile/partition_section_test.rb]: It was in this PR before adding the latest commit

File: test/y2storage/autoinst_profile/partition_section_test.rb [hunk 6]
```diff
@@ -590,15 +590,35 @@ def section_for(name)
...
+        let(:scenario) { "encrypted_random_swap.xml" }
+
+        before do
+          allow_any_instance_of(Y2Storage::Encryption).to receive(:method).and_return(nil)
+        end
+
+        it "does not initialize encryption related elements" do
```
[the remark this answers is not in the document]
Comment (ancorgs) [reply test/y2storage/autoinst_profile/partition_section_test.rb]: I feel that cloning as LUKS1 a device that was in fact using plain encryption, or LUKS2 or any other unsupported system is safer than cloning it as a plain device.

[the remark this answers is not in the document]
Comment (imobachgs) [reply test/y2storage/autoinst_profile/partition_section_test.rb]: I am still not convinced with any of the approaches. But we will clone to `:luks1`.

[the remark this answers is not in the document]
Comment (imobachgs) [reply test/y2storage/autoinst_profile/partition_section_test.rb]: Actually, if we are thinking about security, we should not export the `crypt_key`. Thanks to the changes in this PR, AutoYaST will complain about the missing password and the user will be forced to have a look at the profile.

**Pull request merged**